include/pybind11/pybind11.h

```diff
@@ -403,7 +408,14 @@ class cpp_function : public function {
         strdup_guard guarded_strdup;
 
         /* Create copies of all referenced C-style strings */
-        rec->name = guarded_strdup(rec->name ? rec->name : "");
+        if (rec->name == nullptr) {
+            rec->name = guarded_strdup("");
+        } else if (std::strcmp(rec->name, "@__setstate__") == 0) {
```

Comment:
Is the `@` name following some precedence? If not I would suggest using a more descriptive name, e.g., introducing a namespace and some sort of binding protocol version tags, like, `pybind:2:__setstate__`?

Reply:
After turning this over in my mind I settled on `__setstate__[non-constructor]`.

The ideas behind it are:

* Expressive, as you suggested (I copy-pasted "non-constructor" from one of your comments).
* It's immediately obvious that this cannot possibly be a method name that appears in the Python API. IOW it's a strong hint that it's special.
* It looks a bit like a `typing` annotation (good or bad?).
* I feel it's better not to include "pybind" (binding systems come and go, in the long run, conventions stays because of client code relying on them).